tests/integration/src/typehints.rs

```diff
@@ -0,0 +1,681 @@
+
+pub fn integrate(module: &mut Module) {
+    let i_foo = module.add_interface(make_i_foo_interface());
+    let foo_class = module.add_class(make_foo_class(i_foo.clone()));
+    module.add_class(make_b_class(foo_class.clone(), i_foo.clone()));
+    module.add_class(make_foo_handler());
+    module.add_class(make_arg_typehint_class());
+    module.add_class(make_return_typehint_class());
+    module.add_class(make_arg_default_value_class());
+    module
+        .add_function("integration_function_typehints", |_| phper::ok(()))
+        .argument(
+            Argument::by_val("s")
+                .with_type_hint(ArgumentTypeHint::String)
+                .with_default_value(CString::new("'foobarbaz'").unwrap()),
+        )
+        .argument(
+            Argument::by_val("i")
+                .with_type_hint(ArgumentTypeHint::Int)
+                .with_default_value(CString::new("42").unwrap()),
+        )
+        .argument(
+            Argument::by_val("f")
+                .with_type_hint(ArgumentTypeHint::Float)
+                .with_default_value(CString::new("7.89").unwrap()),
+        )
+        .argument(
+            Argument::by_val("b")
+                .with_type_hint(ArgumentTypeHint::Bool)
+                .with_default_value(CString::new("true").unwrap()),
+        )
+        .argument(
+            Argument::by_val("a")
+                .with_type_hint(ArgumentTypeHint::Array)
+                .with_default_value(CString::new("['a'=>'b']").unwrap()),
+        )
+        .argument(
+            Argument::by_val("m")
+                .with_type_hint(ArgumentTypeHint::Mixed)
+                .with_default_value(CString::new("1.23").unwrap()),
+        )
+        .return_type(ReturnType::by_val(ReturnTypeHint::Void));
+}
+
+fn make_i_foo_interface() -> InterfaceEntity {
+    let mut interface = InterfaceEntity::new(r"IntegrationTest\TypeHints\IFoo");
+    interface
+        .add_method("getValue")
+        .return_type(ReturnType::by_val(ReturnTypeHint::String));
+    interface
+        .add_method("setValue")
+        .argument(Argument::by_val("value"));
+
+    interface
+}
+
+fn make_foo_handler() -> ClassEntity<()> {
+    let mut class = ClassEntity::new(r"IntegrationTest\TypeHints\FooHandler");
+
+    class
+        .add_method("handle", Visibility::Public, |_, arguments| {
+            phper::ok(arguments[0].clone())
+        })
+        .argument(
+            Argument::by_val("foo")
+                .with_type_hint(ArgumentTypeHint::ClassEntry(String::from(I_FOO))),
+        )
+        .return_type(ReturnType::by_val(ReturnTypeHint::ClassEntry(
+            String::from(I_FOO),
+        )));
+
+    class
+}
+
+fn make_foo_class(i_foo: Interface) -> ClassEntity<()> {
+    let mut class = ClassEntity::new(r"IntegrationTest\TypeHints\Foo");
+
+    // leak Interface so that ClassEntry can be retrieved later, during module
+    // startup
+    let i_foo_copy: &'static Interface = Box::leak(Box::new(i_foo));
```

Comment:
I am not happy with this, it's ugly. I wonder whether the whole late-loading-via-functions issue could be removed if we just accepted a String for interface name, and looked it up from globals when needed?

Reply:
Actually, you can find the interface through this method: https://docs.rs/phper/latest/phper/classes/struct.ClassEntry.html#method.from_globals  

I think we can modify the prototype of `implements`, something like:  
```rust
fn implements(&mut self, interface: Interface)  
```  
Then we can think about how to design it better later.

Reply:
Created an issue to address this in a separate PR